data-pipeline/src/concentrator/mod.rs

```diff
@@ -0,0 +1,1158 @@
+
+mod aggregation;
+
+/// Return a Duration between t and the unix epoch
+/// If t is before the unix epoch return 0
+fn system_time_to_unix_duration(t: SystemTime) -> Duration {
+    match t.duration_since(time::UNIX_EPOCH) {
+        Err(_) => Duration::from_nanos(0),
+        Ok(d) => d,
+    }
+}
+
+/// Align a timestamp on the start of a bucket
+#[inline]
+fn align_timestamp(t: u64, bucket_size: u64) -> u64 {
+    t - (t % bucket_size)
+}
+
+/// Return true if the span has a span.kind that is eligible for stats computation
+fn compute_stats_for_span_kind(span: &pb::Span) -> bool {
+    span.meta.get("span.kind").is_some_and(|span_kind| {
+        matches!(
+            span_kind.to_lowercase().as_str(),
+            "server" | "consumer" | "client" | "producer"
+        )
+    })
+}
+
+/// The concentrator compute stats on span aggregated by time and span attributes
+///
+/// The ingested spans are only aggregated if they are root, top-level, measured or if their
+/// `span.kind` is eligible and the `compute_stats_by_span_kind` is enabled.
+#[derive(Debug)]
+pub struct Concentrator {
+    /// Size of the time buckets used for aggregation in nanos
+    bucket_size: u64,
+    buckets: HashMap<u64, StatsBucket>,
+    /// Timestamp of the oldest time bucket for which we allow data.
+    /// Any ingested stats older than it get added to this bucket.
+    oldest_timestamp: u64,
+    /// bufferLen is the number of 10s stats bucket we keep in memory before flushing them.
+    /// It means that we can compute stats only for the last `bufferLen * bsize` and that we
+    /// wait such time before flushing the stats.
+    /// This only applies to past buckets. Stats buckets in the future are allowed with no
+    /// restriction.
+    buffer_len: usize,
+    /// flag to enable aggregation of peer tags
+    peer_tags_aggregation: bool,
+    /// flag to enable computation of stats through checking the span.kind field
+    compute_stats_by_span_kind: bool,
+    /// keys for supplementary tags that describe peer.service entities
+    peer_tag_keys: Vec<String>,
+}
+
+impl Concentrator {
+    /// Return a new concentrator with the given parameter
+    /// - `bucket_size`
+    pub fn new(
+        bucket_size: Duration,
+        now: SystemTime,
+        peer_tags_aggregation: bool,
+        compute_stats_by_span_kind: bool,
+        peer_tag_keys: Vec<String>,
+    ) -> Concentrator {
+        Concentrator {
+            bucket_size: bucket_size.as_nanos() as u64,
+            buckets: HashMap::new(),
+            oldest_timestamp: align_timestamp(
+                system_time_to_unix_duration(now).as_nanos() as u64,
+                bucket_size.as_nanos() as u64,
+            ),
+            buffer_len: 2,
+            peer_tags_aggregation,
+            compute_stats_by_span_kind,
+            peer_tag_keys,
+        }
+    }
+
+    pub fn add_span(&mut self, span: &pb::Span) -> Result<()> {
+        if !(trace_utils::has_top_level(span)
```

Comment:
minor: I think this if statement is complex enough that it should be moved into a separate function for readability / maintainability purposes. Something along the lines of:
```rust
fn should_ignore_span(span: &Span, compute_stats_by_span_kind: bool) -> bool {
    !(trace_utils::has_top_level(span)
        || trace_utils::is_measured(span)
        || (self.compute_stats_by_span_kind && compute_stats_for_span_kind))
        || trace_utils::is_partial_snapshot(span)
}